### PR TITLE
Optipng

### PR DESCRIPTION
**Pull request by surma**
This PR adds support for an OptiPNG based PNG compressor.

Things that need to be done after this lands:

- Return a proper version from the optipng module
- Try to refactor the C code to _not_ rely on Emscripten’s emulated filesystem (use `fmemopen` if at all possible)
- Add zopflipng

### CODE REVIEW

**Comment by jakearchibald** [comment]
I ran a few images through this, then imageoptim to see what extra things it could do:

Paletted 7k to 6k (Zopfli got the best number, but advpng also improved upon the 7k).
Paletted 47k to 44k (Advpng+Zopfli got the best number).
Non-paletted 148k to 142k (Advpng+Zopfli got the best number).

It's notable how fast advpng is and appears to beat the numbers optipng gives us. Zopfli isn't fast, but it gets the numbers.

That said, optipng is pretty close in all cases, so I'm happy for this to land. For "v2", I'd like to explore how much work it would be to port advpng's algorithms to something that works on data in memory. But maybe that's a world of pain.

File: codecs/optipng/README.md [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+## API
+
+### `int version()`
+
+Returns the version of optipng as a number. va.b.c is encoded as 0x0a0b0c
+
+### `ArrayBuffer compress(std::string buffer, {level})`;
+
+`compress` will re-compress the given PNG image via `buffer`. `level` is a number between 0 and 7.
```
Comment (jakearchibald) [inline codecs/optipng/README.md]: Lazy question: are there other options to expose, or is that all it offers?

Comment (surma) [reply codecs/optipng/README.md]: There is some more options, just wanted to get this off my table, honestly:

```
Synopsis:
    optipng [options] files ...
Files:
    Image files of type: PNG, BMP, GIF, PNM or TIFF
Basic options:
    -?, -h, -help	show this help
    -o <level>		optimization level (0-7)		[default: 2]
    -v			run in verbose mode / show copyright and version info
General options:
    -backup, -keep	keep a backup of the modified files
    -clobber		overwrite existing files
    -fix		enable error recovery
    -force		enforce writing of a new output file
    -preserve		preserve file attributes if possible
    -quiet, -silent	run in quiet mode
    -simulate		run in simulation mode
    -out <file>		write output file to <file>
    -dir <directory>	write output file(s) to <directory>
    -log <file>		log messages to <file>
    --			stop option switch parsing
Optimization options:
    -f <filters>	PNG delta filters (0-5)			[default: 0,5]
    -i <type>		PNG interlace type (0-1)
    -zc <levels>	zlib compression levels (1-9)		[default: 9]
    -zm <levels>	zlib memory levels (1-9)		[default: 8]
    -zs <strategies>	zlib compression strategies (0-3)	[default: 0-3]
    -zw <size>		zlib window size (256,512,1k,2k,4k,8k,16k,32k)
    -full		produce a full report on IDAT (might reduce speed)
    -nb			no bit depth reduction
    -nc			no color type reduction
    -np			no palette reduction
    -nx			no reductions
    -nz			no IDAT recoding
Editing options:
    -snip		cut one image out of multi-image or animation files
    -strip <objects>	strip metadata objects (e.g. "all")
Optimization levels:
    -o0		<=>	-o1 -nx -nz				(0 or 1 trials)
    -o1		<=>	-zc9 -zm8 -zs0 -f0			(1 trial)
    		(or...)	-zc9 -zm8 -zs1 -f5			(1 trial)
    -o2		<=>	-zc9 -zm8 -zs0-3 -f0,5			(8 trials)
    -o3		<=>	-zc9 -zm8-9 -zs0-3 -f0,5		(16 trials)
    -o4		<=>	-zc9 -zm8 -zs0-3 -f0-5			(24 trials)
    -o5		<=>	-zc9 -zm8-9 -zs0-3 -f0-5		(48 trials)
    -o6		<=>	-zc1-9 -zm8 -zs0-3 -f0-5		(120 trials)
    -o7		<=>	-zc1-9 -zm8-9 -zs0-3 -f0-5		(240 trials)
    -o7 -zm1-9	<=>	-zc1-9 -zm1-9 -zs0-3 -f0-5		(1080 trials)
Notes:
    The combination for -o1 is chosen heuristically.
    Exhaustive combinations such as "-o7 -zm1-9" are not generally recommended.
Examples:
    optipng file.png						(default speed)
    optipng -o5 file.png					(slow)
    optipng -o7 file.png					(very slow)
```

Comment (jakearchibald) [reply codecs/optipng/README.md]: No problem. We'll create an issue to explore these.

**Comment by jakearchibald** [comment]
Wasn't a criticism, just wanted to document the kind of numbers we're dealing with. Optipng is absolutely good enough for v1.

File: codecs/optipng/build.sh [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    --std=c99 -c \
+    node_modules/optipng/src/opngreduc/*.c \
+    node_modules/optipng/src/pngxtern/*.c \
+    node_modules/optipng/src/gifread/*.c \
+    node_modules/optipng/src/minitiff/*.c \
+    node_modules/optipng/src/pnmio/*.c \
+    node_modules/optipng/src/optipng/*.c
+
+  emcc \
+    --bind -o3 \
```
Comment (jakearchibald) [inline codecs/optipng/build.sh]: Should this be `-O3` (same below)? The build seems to change when I capitalise this.

File: src/codecs/optipng/Encoder.worker.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+          resolve(m);
+        },
+      });
+    });
+  }
+
+  async compress(data: BufferSource, opts: EncodeOptions): Promise<ArrayBuffer> {
+    const m = await this.emscriptenModule;
+    const result = m.compress(data, opts);
+    m.free_result();
```
Comment (jakearchibald) [inline src/codecs/optipng/Encoder.worker.ts]: If it safe to call this before copying the bytes from `result`?

File: src/lib/util.ts [hunk 4]
```diff
@@ -109,7 +109,7 @@ export function blobToArrayBuffer(blob: Blob): Promise<ArrayBuffer> {
   return new Promise((resolve) => {
     const fileReader = new FileReader();
     fileReader.addEventListener('load', () => {
-      resolve(fileReader.result);
+      resolve(fileReader.result! as ArrayBuffer);
```
Comment (jakearchibald) [inline src/lib/util.ts]: What's this bit doing?

File: src/codecs/optipng/options.tsx [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+      level: inputFieldValueAsNumber(form.level),
+    };
+    this.props.onChange(options);
+  }
+
+  render({ options }: Props) {
+    return (
+      <form>
+        <label>
+          Level:
```
Comment (jakearchibald) [inline src/codecs/optipng/options.tsx]: Sorry to bikeshed, but should we go for "Effort" here to match other codecs?

File: codecs/optipng/build.sh (hunk 2 above)
Comment (surma) [reply codecs/optipng/build.sh]: oh yeah, vi screwed my over here I think. Good catch

File: src/codecs/optipng/Encoder.worker.ts (hunk 3 above)
Comment (surma) [reply src/codecs/optipng/Encoder.worker.ts]: Yeah, but I should probably do it afterwards anyways.

File: src/codecs/optipng/options.tsx (hunk 5 above)
Comment (surma) [reply src/codecs/optipng/options.tsx]: Sure

File: codecs/optipng/build.sh (hunk 2 above)
Comment (surma) [reply codecs/optipng/build.sh]: Super good catch! Many bytes saved.

Speaking of, we should probably switch from `-O3` to `-Os` on all codecs. Size is currently more important than peak performance imo.

Comment (surma) [reply codecs/optipng/build.sh]: OptiPNG (JS+wasm) is now 155k gzip’d

File: src/lib/util.ts (hunk 4 above)
Comment (surma) [reply src/lib/util.ts]: Since TypeScript v3, result can be a string, null or an ArrayBuffer.

Comment (surma) [reply src/lib/util.ts]: I’ll add some handling

**Comment by surma** [comment]
@jakearchibald PTAL

**Comment by jakearchibald** [comment]
Yay! Great to see this land given how much of a PITA it was.

File: codecs/optipng/README.md (hunk 1 above)
Comment (jakearchibald) [reply codecs/optipng/README.md]: https://github.com/GoogleChromeLabs/squoosh/issues/159

**Pull request merged**